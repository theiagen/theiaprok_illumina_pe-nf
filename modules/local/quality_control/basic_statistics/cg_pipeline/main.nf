process CG_PIPELINE {
    tag "$meta.id"
    label 'process_single'

    container "us-docker.pkg.dev/general-theiagen/staphb/lyveset:1.1.4f"

    input:
    tuple val(meta), path(reads)
    val genome_length
    val cg_pipe_opts

    output:
    tuple val(meta), path('*_readMetrics.tsv'), emit: cg_pipeline_report
    tuple val(meta), path('DATE.txt'), emit: pipeline_date
    tuple val(meta), path('R1_MEAN_Q.txt'), emit: r1_mean_q
    tuple val(meta), path('R2_MEAN_Q.txt'), emit: r2_mean_q
    tuple val(meta), path('COMBINED_MEAN_Q.txt'), emit: combined_mean_q
    tuple val(meta), path('R1_MEAN_LENGTH.txt'), emit: r1_mean_readlength
    tuple val(meta), path('R2_MEAN_LENGTH.txt'), emit: r2_mean_readlength
    tuple val(meta), path('COMBINED_MEAN_LENGTH.txt'), emit: combined_mean_readlength
    tuple val(meta), path('EST_COVERAGE.txt'), emit: est_coverage
    path "versions.yml", emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    def args = task.ext.args ?: ""
    def prefix = task.ext.prefix ?: "${meta.id}"
    def read1 = reads[0]
    def read2 = reads[1]
    def pipe_opts = cg_pipe_opts ?: "--fast"

    """
    # date and version control
    date | tee DATE.txt

    run_assembly_readMetrics.pl ${pipe_opts} ${read1} ${read2} -e ${genome_length} ${args} > ${prefix}_readMetrics.tsv

    # repeat for concatenated read file
    # run_assembly_readMetrics.pl extension awareness    
    if [[ "${read1}" == *".gz" ]] ; then
      extension=".gz"
    else
      extension=""
    fi
    
    cat ${read1} ${read2} > ${prefix}_concat.fastq\${extension}
    run_assembly_readMetrics.pl ${pipe_opts} ${prefix}_concat.fastq\${extension} -e ${genome_length} > ${prefix}_concat_readMetrics.tsv
    
    python3 <<EOF
    import csv
    #grab output average quality and coverage scores by column header
    coverage = 0.0
    with open("${prefix}_readMetrics.tsv",'r') as tsv_file:
      tsv_reader = list(csv.DictReader(tsv_file, delimiter="\t"))
      for line in tsv_reader:
        if "${read1}" in line["File"]:
          with open("R1_MEAN_Q.txt", 'wt') as r1_mean_q:
            r1_mean_q.write(line["avgQuality"])
          with open("R1_MEAN_LENGTH.txt", 'wt') as r1_mean_length:
            r1_mean_length.write(line["avgReadLength"])

          # run_assembly_readMetrics can report coverage as '.'
          try:
            coverage = float(line["coverage"])
          except ValueError:
            continue
          print(coverage)
          
        else:
          with open("R2_MEAN_Q.txt", 'wt') as r2_mean_q:
            r2_mean_q.write(line["avgQuality"])
          with open("R2_MEAN_LENGTH.txt", 'wt') as r2_mean_length:
            r2_mean_length.write(line["avgReadLength"])
          # run_assembly_readMetrics can report coverage as '.'
          try:
            coverage += float(line["coverage"])
          except ValueError:
            continue

      with open("EST_COVERAGE.txt", 'wt') as est_coverage:
        est_coverage.write(str(coverage))

    # parse concatenated read metrics
    # grab output average quality and coverage scores by column header
    with open("${prefix}_concat_readMetrics.tsv",'r') as tsv_file_concat:
      tsv_reader_concat = list(csv.DictReader(tsv_file_concat, delimiter="\t"))
      for line in tsv_reader_concat:
        if "${prefix}_concat" in line["File"]:
          with open("COMBINED_MEAN_Q.txt", 'wt') as combined_mean_q:
            combined_mean_q.write(line["avgQuality"])
          with open("COMBINED_MEAN_LENGTH.txt", 'wt') as combined_mean_length:
            combined_mean_length.write(line["avgReadLength"])
    EOF

    # R2_MEAN_Q to make SE workflow work otherwise read_float fails
    if [[ ! -f R2_MEAN_Q.txt ]] ; then
      echo "0.0" > R2_MEAN_Q.txt
    fi
    # same for R2_MEAN_LENGTH
    if [[ ! -f R2_MEAN_LENGTH.txt ]] ; then
      echo "0.0" > R2_MEAN_LENGTH.txt
    fi

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        python: \$(python3 --version | cut -d' ' -f2)
        cg_pipeline: staphb/lyveset:1.1.4f
    END_VERSIONS
    """

    stub:
    def prefix = task.ext.prefix ?: "${meta.id}"
    """
    echo "na" > "${prefix}_readMetrics.tsv"
    echo "na" > "DATE.txt"
    echo "na" > "R1_MEAN_Q.txt"
    echo "na" > "R2_MEAN_Q.txt"
    echo "na" > "COMBINED_MEAN_Q.txt"
    echo "na" > "R1_MEAN_LENGTH.txt"
    echo "na" > "R2_MEAN_LENGTH.txt"
    echo "na" > "COMBINED_MEAN_LENGTH.txt"
    echo "na" > "EST_COVERAGE.txt"
    
    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        python: \$(python3 --version | cut -d' ' -f2)
        cg_pipeline: staphb/lyveset:1.1.4f
    END_VERSIONS
    """
}