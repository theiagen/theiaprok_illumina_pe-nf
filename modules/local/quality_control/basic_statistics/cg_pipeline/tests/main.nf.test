nextflow_process {
    name "Test Process CG_PIPELINE"
    script "../main.nf"
    process "CG_PIPELINE"
    
    test("cg_pipeline test") {
        when {
            process {
                """
                input[0] = Channel.of([
                    [ id:'test_sample' ],
                    [
                    file(params.pipelines_testdata_base_path + "genomics/prokaryotes/bacteroides_fragilis/illumina/fastq/test1_1.fastq.gz", checkIfExists: true),
                    file(params.pipelines_testdata_base_path + "genomics/prokaryotes/bacteroides_fragilis/illumina/fastq/test1_2.fastq.gz", checkIfExists: true)
                    ]
                ])
                input[1] = 5200000 // genome length
                input[2] = "--fast"
                """
            }
        }
        then {
            assertAll(
                { assert process.success },
                {  assert snapshot(
                    process.out.cg_pipeline_report,
                    process.out.pipeline_date,
                    process.out.r1_mean_q,
                    process.out.r2_mean_q,
                    process.out.combined_mean_q,
                    process.out.r1_mean_readlength,
                    process.out.r2_mean_readlength,
                    process.out.combined_mean_readlength,
                    process.out.est_coverage,
                    process.out.versions
                    ).match() }
            )
        }
    }
    
    test("cg_pipeline - stub") {
        options "-stub"
        when {
            process {
                """
                input[0] = [
                    [ id:'test_sample' ],
                    file(params.pipelines_testdata_base_path + "genomics/prokaryotes/bacteroides_fragilis/illumina/fastq/test1_1.fastq.gz", checkIfExists: true),
                    file(params.pipelines_testdata_base_path + "genomics/prokaryotes/bacteroides_fragilis/illumina/fastq/test1_2.fastq.gz", checkIfExists: true)
                ]
                input[1] = 5200000 // genome length
                input[2] = "--fast"
                """
            }
        }
        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }
    }
}