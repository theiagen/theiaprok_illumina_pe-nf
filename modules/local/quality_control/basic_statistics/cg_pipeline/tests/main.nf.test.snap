{
    "cg_pipeline - stub": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "test_sample_readMetrics.tsv:md5,51bb9b62d9be48ac08dfb0ec931632e2"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "DATE.txt:md5,51bb9b62d9be48ac08dfb0ec931632e2"
                    ]
                ],
                "2": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "R1_MEAN_Q.txt:md5,51bb9b62d9be48ac08dfb0ec931632e2"
                    ]
                ],
                "3": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "R2_MEAN_Q.txt:md5,51bb9b62d9be48ac08dfb0ec931632e2"
                    ]
                ],
                "4": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "COMBINED_MEAN_Q.txt:md5,51bb9b62d9be48ac08dfb0ec931632e2"
                    ]
                ],
                "5": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "R1_MEAN_LENGTH.txt:md5,51bb9b62d9be48ac08dfb0ec931632e2"
                    ]
                ],
                "6": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "R2_MEAN_LENGTH.txt:md5,51bb9b62d9be48ac08dfb0ec931632e2"
                    ]
                ],
                "7": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "COMBINED_MEAN_LENGTH.txt:md5,51bb9b62d9be48ac08dfb0ec931632e2"
                    ]
                ],
                "8": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "EST_COVERAGE.txt:md5,51bb9b62d9be48ac08dfb0ec931632e2"
                    ]
                ],
                "9": [
                    "versions.yml:md5,841c45908dbf4e61b23b03d38835f0b5"
                ],
                "cg_pipeline_report": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "test_sample_readMetrics.tsv:md5,51bb9b62d9be48ac08dfb0ec931632e2"
                    ]
                ],
                "combined_mean_q": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "COMBINED_MEAN_Q.txt:md5,51bb9b62d9be48ac08dfb0ec931632e2"
                    ]
                ],
                "combined_mean_readlength": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "COMBINED_MEAN_LENGTH.txt:md5,51bb9b62d9be48ac08dfb0ec931632e2"
                    ]
                ],
                "est_coverage": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "EST_COVERAGE.txt:md5,51bb9b62d9be48ac08dfb0ec931632e2"
                    ]
                ],
                "pipeline_date": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "DATE.txt:md5,51bb9b62d9be48ac08dfb0ec931632e2"
                    ]
                ],
                "r1_mean_q": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "R1_MEAN_Q.txt:md5,51bb9b62d9be48ac08dfb0ec931632e2"
                    ]
                ],
                "r1_mean_readlength": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "R1_MEAN_LENGTH.txt:md5,51bb9b62d9be48ac08dfb0ec931632e2"
                    ]
                ],
                "r2_mean_q": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "R2_MEAN_Q.txt:md5,51bb9b62d9be48ac08dfb0ec931632e2"
                    ]
                ],
                "r2_mean_readlength": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "R2_MEAN_LENGTH.txt:md5,51bb9b62d9be48ac08dfb0ec931632e2"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,841c45908dbf4e61b23b03d38835f0b5"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.8"
        },
        "timestamp": "2025-06-23T21:55:30.727040169"
    },
    "cg_pipeline test": {
        "content": [
            [
                [
                    {
                        "id": "test_sample"
                    },
                    [
                        "test_sample_concat_readMetrics.tsv:md5,0e774be52ea4f79886590883f874187a",
                        "test_sample_readMetrics.tsv:md5,f84e51059668e860f12a566258e562c1"
                    ]
                ]
            ],
            [
                [
                    {
                        "id": "test_sample"
                    },
                    "R1_MEAN_Q.txt:md5,1f15e0f43349abcd2427df06d3bb3f18"
                ]
            ],
            [
                [
                    {
                        "id": "test_sample"
                    },
                    "R2_MEAN_Q.txt:md5,ea65ac714a8b6e5c09fc31b5f31c936d"
                ]
            ],
            [
                [
                    {
                        "id": "test_sample"
                    },
                    "COMBINED_MEAN_Q.txt:md5,65f65e7ee6a3b64dabcc495951893e30"
                ]
            ],
            [
                [
                    {
                        "id": "test_sample"
                    },
                    "R1_MEAN_LENGTH.txt:md5,855d0b2f08e1e39e7b9daa1db6ac9716"
                ]
            ],
            [
                [
                    {
                        "id": "test_sample"
                    },
                    "R2_MEAN_LENGTH.txt:md5,855d0b2f08e1e39e7b9daa1db6ac9716"
                ]
            ],
            [
                [
                    {
                        "id": "test_sample"
                    },
                    "COMBINED_MEAN_LENGTH.txt:md5,855d0b2f08e1e39e7b9daa1db6ac9716"
                ]
            ],
            [
                [
                    {
                        "id": "test_sample"
                    },
                    "EST_COVERAGE.txt:md5,b5946ac784beee15be7753b5a581fde6"
                ]
            ],
            [
                "versions.yml:md5,841c45908dbf4e61b23b03d38835f0b5"
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.8"
        },
        "timestamp": "2025-06-23T21:55:24.934480746"
    }
}